### PR TITLE
User login process completed

### PR DESCRIPTION
**Pull request by BenMatase**

### CODE REVIEW

File: backend/api.py [hunk 1]
```diff
@@ -83,13 +98,66 @@ def modify():
...
+        google_id = validate_user(request.values.get('idtoken'))
+    except UserDeniedException as e:
+        # TODO: log failure to access
+        print(e)
+        # return empty response to signify user not given permission
+        return ''
+
     return construct_profile_json(google_id)
 
+<<<<<<< HEAD
```
Comment (BenMatase) [inline backend/api.py]: Solve this conflict that somehow got in

File: backend/api.py [hunk 2]
```diff
@@ -264,73 +332,138 @@ def build_condition(column_name, list_of_vals):
...
 
 def exec_query(sql_string):
-    cursor = mysql.get_db().cursor()
-    try:
-        # Execute the SQL command, by calling execute like this, it handles sql injection
-        cursor.execute(sql_string)
-    except Exception as e:
-        print(e)
-
+    cursor = conn.cursor()
```
Comment (BenMatase) [inline backend/api.py:331]: Should this be once and keep alive or should get new cursor everytime?

File: backend/api.py [hunk 3]
```diff
@@ -264,73 +332,138 @@ def build_condition(column_name, list_of_vals):
...
+
+    return googleid
+
+def send_approval_email(name, email, userid):
+    # TODO: find a better way to do this
+    link = 'http://eg.bucknell.edu/~bdm015/Flower/web/requests/approveUser.html?userid={}'.format(
+        userid
+    )
+
+    body = (EMAIL_HTML_START +
```
Comment (BenMatase) [inline backend/api.py]: Clean this up

File: backend/api.py [hunk 4]
```diff
@@ -264,73 +332,138 @@ def build_condition(column_name, list_of_vals):
...
+                email,
+                link
+            )
+
+    # get list of admin emails
+    get_admins_sql = 'SELECT email FROM user WHERE isAdmin = 1'
+    admins = exec_query(get_admins_sql)
+    admin_emails = [admin['email'] for admin in admins]
+
+    sendEmail('energyhill@bucknell.edu',
```
Comment (BenMatase) [inline backend/api.py]: Do the explicit param naming

File: backend/api.py [hunk 5]
```diff
@@ -264,73 +332,138 @@ def build_condition(column_name, list_of_vals):
...
+              body,
+              [],
+              True
+    )
+
+def send_approved_email(approved_user_email):
+    # TODO: add actual link to dashboard
+    link = ''
+
+    body = (EMAIL_HTML_START +
```
Comment (BenMatase) [inline backend/api.py]: Clean up

File: backend/api.py [hunk 6]
```diff
@@ -264,73 +332,138 @@ def build_condition(column_name, list_of_vals):
...
+            'Hello,\n' +
+            'You have been granted access to the Energy Hill dashboard.  You can use this link to access the dashboard <a href="{0}">{0}</a>' +
+            'Thanks,\n' +
+            'Energy Hill Robot' +
+            EMAIL_HTML_END).format(
+                  link 
+            )
+
+    sendEmail('energyhill@bucknell.edu',
```
Comment (BenMatase) [inline backend/api.py]: Explicity param naming to make more readable

File: backend/api.py (hunk 3 above)
Comment (BenMatase) [reply backend/api.py]: Also make it a good looking email

File: web/control-panel/login.html [hunk 7]
```diff
@@ -21,19 +21,19 @@
...
+          xhr.open('POST', 'http://linuxremote1.bucknell.edu:5001/get-profile');
           xhr.withCredentials = true;
           //xhr.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded');
           xhr.onload = function() {
               console.log('Signed in as: ' + xhr.responseText);
+              
+              // redirect if user is not allowed
+              if (xhr.responseText === '') {
+                // TODO: figure out how to handle https connection is not private issue
+                window.location.replace("http://eg.bucknell.edu/~bdm015/Flower/web/requests/requestApproval.html");
```
Comment (BenMatase) [inline web/control-panel/login.html]: Make this url more dynamic?

File: web/control-panel/login.html [hunk 8]
```diff
@@ -21,19 +21,19 @@
           var id_token = googleUser.getAuthResponse().id_token;
 
           var xhr = new XMLHttpRequest();
-          xhr.open('POST', 'http://134.82.9.147:5000/get-profile');
+          xhr.open('POST', 'http://linuxremote1.bucknell.edu:5001/get-profile');
```
Comment (BenMatase) [inline web/control-panel/login.html]: Make this url more dynamic?

File: web/requests/approveUser.html [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  <body>
+      <div class="g-signin2" data-onsuccess="onSignIn"></div>
+      <script src="https://apis.google.com/js/platform.js" async defer></script>
+      <script type="text/javascript" src="../common/queryParams.js" ></script>
+      <script>
+      function onSignIn(googleUser) {
+          var id_token = googleUser.getAuthResponse().id_token;
+
+          var xhr = new XMLHttpRequest();
+          xhr.open('POST', 'http://linuxremote1.bucknell.edu:5001/approve-user');
```
Comment (BenMatase) [inline web/requests/approveUser.html]: Url more dynamic?

File: web/requests/requestApproval.html [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    <meta name="google-signin-client_id" content="438120227370-65o4j0kmk9mbij1bp3dqvnts9dh4kfrf.apps.googleusercontent.com">
+
+    <meta charset="utf-8">
+  </head>
+
+  <style>
+  </style>
+
+  <body>
+      <h1>Looks like you don't have access to the dashboard.  If you think this is a mistake, click the button below.</h1>
```
Comment (BenMatase) [inline web/requests/requestApproval.html:17]: Make this prettier in another task.  Probably later

Comment (BenMatase) [reply web/requests/requestApproval.html:17]: Also make this actually a button once we get sessions

File: web/requests/requestApproval.html [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+      <div class="g-signin2" data-onsuccess="onSignIn"></div>
+      <script src="https://apis.google.com/js/platform.js" async defer></script>
+      <script>
+      function onSignIn(googleUser) {
+          var id_token = googleUser.getAuthResponse().id_token;
+
+          var xhr = new XMLHttpRequest();
+          xhr.open('POST', 'http://linuxremote1.bucknell.edu:5001/request-access');
```
Comment (BenMatase) [inline web/requests/requestApproval.html]: More dynamic URL?

File: backend/api.py (hunk 2 above)
Comment (BenMatase) [reply backend/api.py:331]: According to the internet, we don't need to close cursors for mysqldb cursors.. so I'm not worrying about it

File: web/control-panel/login.html (hunk 7 above)
Comment (BenMatase) [reply web/control-panel/login.html]: Leaving for now, will talk to Marchiori as of issue #279

File: backend/alerts.py [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+    cursor = db.cursor()
+    sql = "SELECT * FROM alerts WHERE handled != 1;"
+
+    try:
+        # Execute the SQL command
+        cursor.execute(sql)
+        # Fetch all the rows in a list of lists.
+        unhandled = cursor.fetchall()
+    except:
+        print("Error: unable to fecth data")
```
Comment (DTV96Calibre) [inline backend/alerts.py:32]: Fecth is a misspelling

File: backend/api.py [hunk 13]
```diff
@@ -52,7 +56,13 @@ def get():
 # TODO: need to commit after executing insert
```
Comment (DTV96Calibre) [inline backend/api.py]: There are a butt ton of todos mostly related to error logging. Maybe address these before we start needing to identify what's wrong.

File: backend/api.py [hunk 14]
```diff
@@ -52,7 +56,13 @@ def get():
...
 @app.route('/insert')
 def insert():
-    validate_user(request.values.get('id_token'))
+    try:
+        validate_user(request.values.get('id_token'))
+    except UserDeniedException as e:
```
Comment (DTV96Calibre) [inline backend/api.py:61]: This exception will be silenced without any indication that something went wrong.

Comment (DTV96Calibre) [reply backend/api.py:61]: This is repeated below

File: web/control-panel/login.html (hunk 8 above)
Comment (DTV96Calibre) [reply web/control-panel/login.html]: We should have these deployment specific values stored in config files.

**Review by DTV96Calibre — CHANGES_REQUESTED** [review]
Error logging should be handled before we start using this code.

**Review by DTV96Calibre — APPROVED** [review]

**Pull request merged**